src/normalizer/token_classifier.rs

```diff
@@ -0,0 +1,91 @@
+use std::collections::HashSet;
+
+use crate::{Token, TokenKind};
+use crate::token::SeparatorKind;
+use super::Normalizer;
+
+#[derive(Debug, Default, Clone)]
+pub struct TokenClassifier {
+    hard_separators: HashSet<char>,
+    soft_separators: HashSet<char>,
+    stop_words: HashSet<String>,
+}
+
+impl TokenClassifier {
+    pub fn new(stop_words: HashSet<String>, soft_separators: HashSet<char>, hard_separators: HashSet<char>) -> Self {
+        Self { soft_separators, stop_words, hard_separators }
+    }
+}
+
+impl Normalizer for TokenClassifier {
+    fn normalize<'a>(&self, mut token: Token<'a>) -> Token<'a> {
+        let word = token.word.as_ref();
+        let mut is_hard_separator = false;
+        if self.stop_words.contains(word) {
+            token.kind = TokenKind::StopWord;
+            token
+        } else if word.chars().all(|c| {
+            is_hard_separator = self.hard_separators.contains(&c);
+            self.soft_separators.contains(&c) || is_hard_separator
+        }) {
+            if is_hard_separator {
+                token.kind = TokenKind::Separator(SeparatorKind::Hard);
+            } else {
+                token.kind = TokenKind::Separator(SeparatorKind::Soft);
+            }
+            token
```

Comment:
this seems a bit overengineered, suggesting that the tokenizer could return a multicharacter separator.
Moreover, it is no consistent with stop words,  we don't check if any stop word in the list is a subset of the token.
Moreover, if I create a tokenizer which detects smiley like `;-)` as one token, it would be treated as a separator by this normalizer.
I would just compare the provided token with the list of separators.